File: app/src/domain/calculations/customerGap.ts
/**
 * 前年比客数GAP 計算
 *
 * 客数変動では説明できない、1人あたり購買行動の変化を数値化する。
 *
 * - 点数客数GAP = 点数前年比 − 客数前年比
 * - 金額客数GAP = 金額前年比 − 客数前年比
 *
 * 例: 客数100%、点数95% → 点数客数GAP = ▲5%
 *   「客数は同じだが、1人あたりの買上点数が5%減少した」
 *
 * @see references/01-principles/customer-gap-definition.md
 * @see references/01-principles/calculation-canonicalization-map.md
 */
import { z } from 'zod'
import { safeDivide } from './utils'

// ── 入力契約 ──

export interface CustomerGapInput {
  /** 当期客数 */
  readonly curCustomers: number
  /** 前期客数 */
  readonly prevCustomers: number
  /** 当期販売点数 */
  readonly curQuantity: number
  /** 前期販売点数 */
  readonly prevQuantity: number
  /** 当期売上金額 */
  readonly curSales: number
  /** 前期売上金額 */
  readonly prevSales: number
}

export const CustomerGapInputSchema = z.object({
  curCustomers: z.number(),
  prevCustomers: z.number(),
  curQuantity: z.number(),
  prevQuantity: z.number(),
  curSales: z.number(),
  prevSales: z.number(),
})

// ── 出力契約 ──

export interface CustomerGapResult {
  /** 客数前年比（小数: 1.0 = 100%） */
  readonly customerYoY: number
  /** 点数前年比（小数: 1.0 = 100%） */
  readonly quantityYoY: number
  /** 金額前年比（小数: 1.0 = 100%） */
  readonly salesYoY: number
  /** 点数客数GAP = 点数前年比 − 客数前年比（小数: -0.05 = ▲5%） */
  readonly quantityCustomerGap: number
  /** 金額客数GAP = 金額前年比 − 客数前年比（小数: -0.05 = ▲5%） */
  readonly amountCustomerGap: number
}

export const CustomerGapResultSchema = z.object({
  customerYoY: z.number(),
  quantityYoY: z.number(),
  salesYoY: z.number(),
  quantityCustomerGap: z.number(),
  amountCustomerGap: z.number(),
})

// ── 計算関数 ──

/**
 * 前年比客数GAPを計算する。
 *
 * 前期の客数・点数・金額のいずれかが 0 の場合は null を返す
 * （0除算防止 + 意味のない GAP 値を表示しない）。
 */
export function calculateCustomerGap(input: CustomerGapInput): CustomerGapResult | null {
  const { curCustomers, prevCustomers, curQuantity, prevQuantity, curSales, prevSales } = input

  // 前期データが不足 → 計算不可
  if (prevCustomers <= 0 || prevQuantity <= 0 || prevSales <= 0) {
    return null
  }

  const customerYoY = safeDivide(curCustomers, prevCustomers, 0)
  const quantityYoY = safeDivide(curQuantity, prevQuantity, 0)
  const salesYoY = safeDivide(curSales, prevSales, 0)

  return {
    customerYoY,
    quantityYoY,
    salesYoY,
    quantityCustomerGap: quantityYoY - customerYoY,
    amountCustomerGap: salesYoY - customerYoY,
  }
}

File: wasm/customer-gap/src/customer_gap.rs
/// Customer Gap calculation module.
///
/// Calculates YoY ratios and gap metrics for customer count, quantity, and sales.
/// Gap = specific YoY ratio - customer YoY ratio (per-capita behavior change).
///
/// Returns None if any prior period value <= 0 (invalid comparison base).
///
/// @contractId BIZ-013
/// @semanticClass business
/// @methodFamily behavioral

use crate::utils::safe_divide;

pub struct CustomerGapResult {
    pub customer_yoy: f64,
    pub quantity_yoy: f64,
    pub sales_yoy: f64,
    pub quantity_customer_gap: f64,
    pub amount_customer_gap: f64,
}

/// Calculate customer gap metrics.
/// Returns None if any prev value <= 0.
pub fn calculate_customer_gap(
    cur_customers: f64,
    prev_customers: f64,
    cur_quantity: f64,
    prev_quantity: f64,
    cur_sales: f64,
    prev_sales: f64,
) -> Option<CustomerGapResult> {
    if prev_customers <= 0.0 || prev_quantity <= 0.0 || prev_sales <= 0.0 {
        return None;
    }

    let customer_yoy = safe_divide(cur_customers, prev_customers, 0.0);
    let quantity_yoy = safe_divide(cur_quantity, prev_quantity, 0.0);
    let sales_yoy = safe_divide(cur_sales, prev_sales, 0.0);

    Some(CustomerGapResult {
        customer_yoy,
        quantity_yoy,
        sales_yoy,
        quantity_customer_gap: quantity_yoy - customer_yoy,
        amount_customer_gap: sales_yoy - customer_yoy,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn basic_gap() {
        let r = calculate_customer_gap(220.0, 200.0, 550.0, 500.0, 110_000.0, 100_000.0).unwrap();
        assert_eq!(r.customer_yoy, 1.1);
        assert_eq!(r.quantity_yoy, 1.1);
        assert_eq!(r.sales_yoy, 1.1);
        assert_eq!(r.quantity_customer_gap, 0.0);
        assert_eq!(r.amount_customer_gap, 0.0);
    }

    #[test]
    fn gap_with_difference() {
        let r = calculate_customer_gap(200.0, 200.0, 600.0, 500.0, 130_000.0, 100_000.0).unwrap();
        assert_eq!(r.customer_yoy, 1.0);
        assert_eq!(r.quantity_yoy, 1.2);
        assert!((r.quantity_customer_gap - 0.2).abs() < 1e-10);
        assert!((r.amount_customer_gap - 0.3).abs() < 1e-10);
    }

    #[test]
    fn null_on_zero_prev_customers() {
        assert!(calculate_customer_gap(100.0, 0.0, 100.0, 100.0, 100.0, 100.0).is_none());
    }

    #[test]
    fn null_on_negative_prev() {
        assert!(calculate_customer_gap(100.0, -1.0, 100.0, 100.0, 100.0, 100.0).is_none());
    }

    #[test]
    fn zero_current_values() {
        let r = calculate_customer_gap(0.0, 100.0, 0.0, 100.0, 0.0, 100.0).unwrap();
        assert_eq!(r.customer_yoy, 0.0);
        assert_eq!(r.quantity_yoy, 0.0);
        assert_eq!(r.sales_yoy, 0.0);
    }
}
